Add number ranges to download commands (fixes #2384)

Al-Caveman opened the pull request:
fixes it only as per the suggestion from pkill-nine here [1] by adding the
argument --all/-a to download-delete. but does not add the fancy selection
thingy that i asked for in #2384.

[1] https://github.com/qutebrowser/qutebrowser/issues/2384#issuecomment-283448617

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/2493)
<!-- Reviewable:end -->

The-Compiler commented:
Like mentioned in IRC, I don't think it makes sense to add `--all` when we are going to add ranges later.

The-Compiler commented:
What about moving the "get downloads based on a range" logic to a method, and then also changing the other `:download-*` commands to take such ranges for consistency?

Al-Caveman commented:
I agree. I think I did a very dirty job. I should re-think the whole thing, and rewrite as a more generic method  (as you said) that is reused elsewhere (instead of introducing my exceptional cases here and there).

The-Compiler commented:
I don't think `_download_select` is an usful abstraction to have, at least as-is - instead of making the calling code easier, it makes it much more complicated.

I think either the caller should do the filtering like before, or maybe it'd make sense to pass a function to `_download_select`, roughly like this:

```python
def _download_select(self, condition, errormsg, count, indexset):
    downloads = ...
    for d in downloads:
        if not condition(d):
            raise cmdexc.CommandError("Download {} {}".format(d, errormsg))

downloads = self._download_select(lambda d: d.done, "is not done", ...)
```

I also think `nonexistent` can just fail by default in `_download_select` already. With that, `_raise_error_download` probably isn't needed anymore either.

`--all` should also continue to work for now. You can use `message.warning` to show a warning about it being deprecated though, and add `@cmdutils.argument('all_', hide=True)` to hide it from the docs. See `def download` in `browser/commands.py` for an example.

---

As for tests, the first step would be to keep the existing ones running - a lot of them [fail](https://travis-ci.org/qutebrowser/qutebrowser/jobs/220796397) currently. I'd really suggest running the tests locally (at least `tests/end2end/features/test_downloads_bdd.py`) - see [the CONTRIBUTING file](https://github.com/qutebrowser/qutebrowser/blob/master/CONTRIBUTING.asciidoc#running-specific-tests) for details.

After that, it'd good to have tests for various other scenarios:

- Unittests for `parse_number_sets` (with a parametrized test) in `tests/unit/utils/test_utils.py`
- Tests for using at least one of the download functions with index sets
- Making sure `--all` still works everywhere

Al-Caveman commented:
Thank you for spending your time to read my code, and to comment on it.

I will think about how to simplify that, plus learn how to use qutebrowser tests. 

For now, cb76a87 adds:

-  `--all/-a` to `download-*`, 
- A new future: when printing errors that need to show lists of numbers, it summarizes the list back to something like "1,2-100" instead of printing the entire list. This list-to-description function is in utils.py.

And 7887913 fixes:
- A formatting fix in some of the errors.

Al-Caveman commented:
This commit a60f28e has everything applied by using lambdas as per your advise. I made only modification, which is that I have also extended lambdas to, not only selection criteria, but also the actions.

Basically, here is an example:

    self._selective_dos(
          count,                                # support count
          all_,                                 # support --all/-a
          indexset,                             # select downloads
          lambda d: d.done,                     # matching this
          [
              lambda d: d.delete(),
              lambda d: d.remove(),
              lambda d: d.remove(),
              lambda d: log.downloads.debug("deleted download {}".format(d))
          ],                                    # to do these against
          "are not successfully downloaded yet" # error postfix
      )

All my changes are only in `downloads.py`.

Kindly feel free to propose more corrections/fixes. One thing I'm thinking about is to learn how to write tests. 

(side note: thank you for your comments; you made me learn lambdas :-p)

The-Compiler commented:
Sorry, but this is getting more and more cryptic. If every line needs a comment, it probably means your code isn't exactly readable. :wink:

NoctuaNivalis commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,196 @@ def _on_data_changed(self, idx, *, webengine):
 
-    def _raise_no_download(self, count):
-        """Raise an exception that the download doesn't exist.
+    def _selective_dos(self, count, all_, indexset, criterion, dos,
```
I'd replace this with a generator method, no?

Al-Caveman replied:
The problem is that, if I delete download `self[5]`, then immediately what was download `self[6]` will get pointed by index 5 instead. I.e. indexes change as we modify downloads. So what was `self[6]` will become `self[5]` if we delete `self[5]`.

This is why I think it's better to identify all downloads:
  1. Matching criteria.
  2. Not matching criteria.
  3. Not existing.

And put them all in their bags, all before we do anything to them, so that once we start doing things to them (e.g. deleting them) and causing their indexes to change, we will not have to figure out "hmm was the index of this download previously (say) 9?", because we have already put them in bags.

Then goes printing error messages. I am aggregating all errors for all downloads, so that I show them in one message. E.g. "Downloads 1,5-7,9,l-100 are not done yet!". Doing such error messages will require me to kinda aggregate all downloads first, and them summarize them.

Of course, I don't think it's impossible to do that with an iterator. But I just don't see how if I do it with an iterator, it will be simpler. 

Maybe I'm wrong. Afterall, this is the 1st time I use a lambda for anything useful. Maybe you know something that I don't. If that's the case, I'd appreciate if you show me how using a generator/iterator would help here.

The-Compiler replied:
I think it's fine to return a list in this case - because (as @Al-Caveman mentioned) at least with the current code, the list of downloads is getting mutated while iterating over the matched downloads.

The-Compiler commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,212 @@ def _on_data_changed(self, idx, *, webengine):
+
+        # delete them
+        for d in matchs:
+            d.delete(),
```
Remove the commas here and on the next line

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
     pass
 
 
+def parse_numsettxt_into_list(txtset, nummax):
```
This still needs unittests :wink:

Al-Caveman replied:
The asymptotic worst run time complexity of this function is O(max(txtset)). E.g. if someone defines txtset = 1,3,5-100000, then max(txtset) = 100000.

Once this potentially long list is generated, other commands will evaluate them one by one (that's how I did them so far). E.g. `download-delete 1,3,5-100000` will cause a loop to be run in `_select_downloads` that will have 100009 iterations. So because of this, the worst run time complexity of `_select_downloads` is also O(max(txtset)).

---

Alternatively, I can change this, so that the asymptotic worst run time becomes O(number of downloads). So if we have, say 50 download objects above the tabs, then we will have 50 iterations.

If I rub more neurons, I can probably get it down to O(min(txtset, number of downloads)).

But the method that is in my mind, while will allow us to change the complexity from O(max(txtset)) down to O(number of downloads), or even O(min(txtset, number of downloads)), would probably be more expensive per each step.

Basically, `parse_numsettxt_into_list` will become `is_number_in_numsettxt(num, txtset)`, and it will return `True` or `False` if number `num` is in the set described by the text `txtset`. This way, if the user defines some crazy set like `1-112342342534`, we will not return a numbers list that has all those crazy numbers.

Or maybe to reduce function calls, `is_numbers_in_numsettxt(numbers, txtset)`, where the function will return two sets of numbers, one that contains numbers that exist in `txtset`, and another that contains those that don't.

----

Or maybe I am missing a much simpler solution.

Kingdread replied:
You can probably save some time (and space!) by using [`itertools.chain`](https://docs.python.org/3.6/library/itertools.html#itertools.chain) (or `chain_from_iterable`) instead of eagerly creating a list with all numbers. `range` is already lazy on Python 3, so you just need to `chain` those ranges together to get a lazy result. You could probably also work with `yield from` instead of `chain`, whichever you prefer (`yield from range(int(start), int(end)+1)` should do the trick).

Also, I don't think it's necessary to squeeze out the *very best* asymptotic runtime. The numbers we're dealing with here are small (`O(number of downloads)` is probably less than 100 in most cases). In this case I think readable code is more important. And if we notice that this function really is a performance bottleneck (I doubt it), we can always profile, benchmark and optimize later.

The-Compiler replied:
Does qutebrowser actually hang because of this when doing something stupid like `:download-delete 1,3,5-100000`? if not: Don't optimize it if it ain't broken :wink:

Al-Caveman replied:
It doesn't freeze for the specific stupid that is `1-100000`, but it does freeze for about a second for this similar stupid `1-1000000` that has only one more zero. Using my ultra-low voltage i7 laptop (i7-5600U CPU @ ~2.6 GHz). 

I'll have a crack at it. It will probably also simplify the reverse function `parse_list_into_numsettxt`.

NoctuaNivalis replied:
Easiest way to optimize this is probably just saving the range-set as just that: a set of ranges. Then you can check whether a download is included in `O(r)`, *r* being the number of ranges (thank you Python 3 for constant time range inclusion checks), As each range had to be typed manually, there can't be much. As we check each download, total would be `O(d*r)`, *d* being the number of downloads.

Saving as a set of ranges would be `set(range(1, 2), range(2, 3), range(4,1000000))` for `1,2,4-999999`.

Al-Caveman replied:
Yeah, I was pretty much thinking of that when I wrote `is_numbers_in_numsettxt`, with the only exception being ordering the ranges, so that (after the ordering) I can apply a small trick so that the worst run time complexity becomes `O(max(d,r))` instead of `O(d*r)`.

Does Python 3 use Timsort? If so, then probably that's the bottleneck now in terms of asymptotic worst time complexity analysis. I think, after considering Timsort's complexity, the overall complexity is `O(max(d log d, r log r))`, which is practically much better than `O(len(possibly-crazy user-defined index set))`.

Now that the problem is Timsort, I can now rest the time complexity case :-p.

I'll shortly commit a version with these changes. I did not use `range(...)` because I wanted to sort the intervals to avoid `O(d*r)` and achieve `O(max(d log d, r log r))` instead. As far as I tried, it seems to me that lists of ranges cannot be sorted :man_shrugging:.

Kingdread replied:
I think we're optimizing the wrong thing here. The thing you want to optimize here is not the total time that it takes to calculate the list, but rather the "response time", which is the time it takes to actually start processing the list. If you do something stupid like `1-100000`, you probably don't want to eagerly create a list of 100000 items, just to notice that you have only 2 downloads in total.

The key here (IMO) is laziness, which can be achieved by using generators instead of lists or smart in-range checks (like Python's `range`). By offloading the total list-creation-time to small steps in between each iteration, we can have a responsive interface with a sub-optimal algorithm, because in many cases we don't even need to create the whole list of all numbers.

This also improves the user interface for those functions. I think a method which returns numbers (such as your first version) is nicer to use than a method for which you have to pass a list of numbers manually. Another nice interface is proper object with a `__contains__` method (like `range`), which would turn

```
ins = utils.which_nums_in_numints(
        range(1, len(self) + 1),
        indexset_p
      )
```

into something like

```
ins = [i for i in range(1, len(self) + 1) if i in indexset_p]
```

At this point you can even iterate over the downloads directly and throw away those that are not in the indexset:

```
for i, download in enumerate(self):
    if i not in indexset_p:
        continue
    ...
```

And if you sacrifice a bit of runtime performance for readability, this `__contains__` method could look like

```
def __contains__(self, item):
    return any(item in r for r in self.ranges)
```

assuming that `self.ranges` is just a list of `range`-objects. That's more readable than a fully-optimized version, while still offering good performance (as long as you don't specify 10000 ranges on the command line).

In the end, it probably depends on the order in which you iterate over the things: do you iterate over all downloads and check if the index is in the indexset, or do you iterate over the indexset and check if there is a corresponding download. The first case is probably nicer, precisely because of cases like `1-1000000`. Other than that, I'd say worrying about runtime complexity is a non-issue here. I doubt that this function will be the bottleneck for anything, and even then, we can still optimize later (after profiling what really takes the most time, as asymptotic runtimes are misleading).

Also, you can sort a list of ranges by using their `.start` attribute, e.g. `list_of_ranges.sort(key=lambda r: r.start)`.

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+                number.
+        nummax: The number that will be used to replace 'last'
+    """
+    txtset = txtset.lower().replace('last', '%d' % (nummax))
```
You can use `str(nummax)` instead of `'%d' % (nummax)`

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+        nummax: The number that will be used to replace 'last'
+    """
+    txtset = txtset.lower().replace('last', '%d' % (nummax))
+    l = set()
```
Call this `numbers` or so instead - `l` is a bad name because it's easily mixed up with `I` depending on what font people are using.

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+    txtset = txtset.lower().replace('last', '%d' % (nummax))
+    l = set()
+    parts = txtset.split(',')
+    for part in parts:
```
No need to have an extra variable for `parts`, IMHO.

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+    l = set()
+    parts = txtset.split(',')
+    for part in parts:
+        if part.find('-') != -1:
```
`if '-' in part:`

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+    parts = txtset.split(',')
+    for part in parts:
+        if part.find('-') != -1:
+            [start, end] = part.split('-')
```
No need for `[` and `]`

Al-Caveman replied:
Should I also remove `(`  and `)` in `downloads.py` where I did things like `(matchs, nomatchs) = self._select_downloads(...)`?

The-Compiler replied:
Yes please!

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+        if part.find('-') != -1:
+            [start, end] = part.split('-')
+            start = int(start)
+            end = int(end)
```
No need for those two lines since you're calling `int()` below anyways.

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -864,6 +864,63 @@ def unused(_arg):
+    return sorted(l)
+
+
+def parse_list_into_numsettxt(l):
```
I'm really not sure if the complexity of this is worth it. I think it'd be better to just return a comma-separated list, even if it's `1,2,3`. Just seems like a minor benefit for something which isn't immediately obvious when looking at the code, so I'm inclined to drop it.

Al-Caveman replied:
Currently, my issue is that, with my previous function `parse_numsettext_into_list(...)`, someone could do things like `:download-delete 1-123443`, and if we go with the list one, then the error will be like "Downloads 5,6,7,8,9,10,...,123443 don't exist" (without "...").

That's why I put this function, to undo that crazy long list into "5-123443" (assuming downloads 1,2,3,4 were deletable).

The asymptotic worst run time complexity of this function is obviously O(len(numbers)). Maybe by itself is not a problem, but it can be a very large number depending on whether someone defines a crazy number set. For example, if someone does `:download-delete 1-1234325`, then the len(numbers) = 1234321 (if 5 got deleted).

The-Compiler replied:
I mean code complexity, not runtime complexity - I can see how it's useful, but it definitely needs unittests too.

Al-Caveman commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,212 @@ def _on_data_changed(self, idx, *, webengine):
     @cmdutils.register(instance='download-model', scope='window')
     @cmdutils.argument('count', count=True)
-    def download_delete(self, count=0):
+    def download_delete(self, indexset=None, all_=False, count=0):
```
My next commit will have --all/-a only for `download_cancel` and `download_remove`. I think those were the only commands that had --all/-a.

I think `:download-clear` should also be deprecated? Since it seems to me that it is equivelent to `:download-remove -a` or `:download-remove 0-last`?

The-Compiler replied:
`:download-clear` used to be gone, but it wasn't obvious to a lot of people that `:download-remove -a` would do the same - so I decided to bring it back, because it was a common issue people had.

The-Compiler commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,212 @@ def _on_data_changed(self, idx, *, webengine):
+                else:
+                    nomatchs.append(i)
+            except IndexError:
+                noexists.append(i)
```
No, it's fine to gather the downloads which didn't exit in a list - but why return that list and show the error outside of `_select_downloads` everywhere instead of doing it here and just returning matching downloads?

Al-Caveman replied:
I see. I agree. My commit did that (after posting that).

The-Compiler commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,212 @@ def _on_data_changed(self, idx, *, webengine):
+
+        # show errors if needed
+        if len(nomatchs):
+            message.error("Downloads {} are already done!".format(
```
What would you need to edit in `_select_downloads` when `download_pause` is added, with my suggestion?

The thing is, if we add some logic to show singular/plural correctly when there's only one download, it'd be much nicer to only have that logic once instead of everywhere where it's called. Also, there's only one clear return value in that case. I think the benefits are more than the drawbacks (showing errors in `_select_downloads`).

Al-Caveman replied:
As for first paragraph: Sorry. I was wrong. I forgot about passing error messages as arguments.

As for 2nd paragraph: There are different singular/plural forms that depend on the error message. For example these use is/are:

  1. Downloads 1,2,3 are successful.
  2. Download 1 is successful.

While these use doesn't/don't:

  1. Downloads 1,2,3 don't exist.
  2. Download 1 doesn't exist.

I guess, I'll also need to add one more argument to `_select_downloads` that specifies the plural/singular type? :-p E.g. `_select_downloads(plural_type='is/are', ...)`

My latest commit moved common errors (about lack of existence) in `_select_downloads`, while keeps the other function-specific errors in the functions, plus some way of singular/plural correctness.

The-Compiler replied:
When the "doesn't exist" error is handled in `_select_downloads`, there only are `is/are` messages left, no?

Al-Caveman replied:
Right. I'll do it. Sorry for my stupidity. I think by now, it would have been cheaper if you had this feature implemented by yourself :-p.

Al-Caveman commented:
Key changes are:

 - Removed the function `parse_list_into_numsettxt`, and replaced it by `','.join([str(i) for i in nomatchs])' in line `969` of `downloads.py`.
 - Gave up on summarizing lists of downloads in error messages:

   - For downloads that don't match the query: the code now just prints a message saying "No downloads with indexes that match '...'!".
   - For downloads that don't match the criteria: we get the usual per-function errors, with all affected downloads listed. I think summarizing them is needless. If a user has 100 downloads, then I'm don't think that he/she will feel sad by seeing a list of 100 in a single error message.

Al-Caveman commented:
Is this fine? :-p could someone please guide me on how to do it.

Kingdread commented on tests/unit/utils/test_utils.py:
```diff
@@ -926,6 +926,21 @@ def test_system_default_application(self, caplog, config_stub, mocker):
+@pytest.mark.parametrize('txtset, nummax, expected', [
+    ('0-5,9-100,last', 500, [(0, 5), (9, 100), (500, 500)]),
+    ('0-5,9-100,last-3', 500, [(0, 5), (9, 100), (500, 3)]),
+    ('0-5,9-100,last-7000', 500, [(0, 5), (9, 100), (500, 7000)]),
```
Some edge cases that are worth testing are missing:

* unordered values: `6-9,3-5`
* overlapping values: `5-7,0-10`
* malformed input: `-5`, `0--3`, `foo-bar`

Al-Caveman replied:
Added the 1st two tests, but didn't add the last one. The last one should raise an exception, which is then handled elsewhere. I don't know how to use `assert` with exceptions. Ideally, I'd like to assert that the 3rd case will result in a value error exception.

Thoughts?

Al-Caveman replied:
My latest comment has my 1st attempt to test for the last one (malformed input).

Kingdread replied:
You can use `pytest.raises`:

```python
with pytest.raises(ValueError):
    utils.parse_...
```

Kingdread commented on tests/unit/utils/test_utils.py:
```diff
@@ -926,6 +926,21 @@ def test_system_default_application(self, caplog, config_stub, mocker):
+    ([3, 6, 50], [(0, 5), (9, 100), (500, 3)], [3, 50]),
+    ([501, 0], [(0, 5), (9, 100), (500, 7000)], [0, 501]),
+])
+def which_nums_in_numints(nums, numints, expected):
```
The `test_` prefix is missing here.

Kingdread commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,203 @@ def _on_data_changed(self, idx, *, webengine):
+        if all_ == True:
+            indexset = '1-last'
+            message.warning(
+                "'--all/-a' will be deprecated. Kindly use '0-last' instead."
```
This line contradicts the indexset above: Should you use `1-last` or `0-last`?

Al-Caveman replied:
Should've been `1-last`. Fixed now. Thank you.

Kingdread commented on qutebrowser/browser/downloads.py:
```diff
@@ -894,110 +894,203 @@ def _on_data_changed(self, idx, *, webengine):
+                  )
+
+        # identify downloads indexes that fall in the indexset
+        ins = utils.which_nums_in_numints(
```
I still believe that `indexset_p` should have its own class (after all, you treat is as an opaque object), and that this method should be a method on this object, either by using `__contains__` for single elements, or calling it `intersection` if you want to check a whole list of numbers at once (Python's `set` also has `.intersection`):

```python
ins = [i for i in range(1, len(self) + 1) if i in indexset_p]
# or
ins = indexset_p.intersection(range(1, len(self) + 1))
```

Al-Caveman replied:
Not sure I understand. Do you mean something like:

    indexset_p = IndexSet('0-3,8-last')
    ins = indexset_p.intersection(range(1,len(self)+1))

?

I guess that class `IndexSet` to be defined in `utils.py`?

Last question, got any better name suggestions than `IndexSet`?

Kingdread replied:
Yeah, that's what I suggest - making it its own class. I think giving the text in the constructor is nicer than having to set the `.text` attribute, so I'd prefer the first way. But you could also consider making a constructor that takes the "parsed" form (like `IndexSet((0, 3), (8, None))`), and the exposing a static method to parse a string (`IndexSet.from_string('0-3,8-last')`).

Depending on how you implement it, `IndexSet` or `NumberSet` is probably fine.

You could think about making it work with arbitrary `set`s (and `set` here means something that supports the `... in ...` check). This means that you could lazily build the union of any `set`s, and you could use the lazy `range` to make it work for indices:

```python
class SetUnion:
    def __init__(self, sets):
        self.sets = sets
    def __contains__(self, other):
        return any(other in s for s in self.sets)

indexset_p = SetUnion([range(0, 3), range(8, 10)])
2 in indexset_p # True
```

Now you have a pretty simple implementation that works for any sets. You lose some performance, as membership testing is now `O(number of sets)` (assuming that the inner sets support testing in constant time, which `range` does), but the code is very clear (and "number of sets" is probably within limits).

Kingdread commented:
The tests seem fine, though when testing, you should focus on the edge cases (what *could* go wrong?). An edge-case test is worth more than repeated valid-case tests, as most of the valid cases take the same code path anyway, and so if one valid input is correct, the others will probably be too (the exception is if you have multiple valid formats).

The-Compiler commented:
Any update on this, @Al-Caveman?

Al-Caveman commented:
@The-Compiler, short answer is: not much :-p, the added feature itself looks fine to me (been using it for a while, although not aggressively). I'm currently (and still) pending at completing the tests. That's the update.

Now, I'm giving the tests another shot, hopefully I'll complete them in a few hours from now. It's the 1st test I do in my entire life (yes I'm this lame).

Also sorry for my late reply, I've just noticed that you wrote this comment here (after 18 days).

Al-Caveman commented:
Any idea how to run the tests? I tried this `tox -e py34 -- tests/unit/utils/test_utils.py` and got this:
 
    py34 installed: beautifulsoup4==4.6.0,cheroot==5.5.0,click==6.7,colorama==0.3.9,coverage==4.4.1,cssutils==1.0.2,decorator==4.0.11,EasyProcess==0.2.3,fields==5.0.0,Flask==0.12.2,glob2==0.5,httpbin==0.5.0,hunter==1.4.1,hypothesis==3.11.1,itsdangerous==0.24,Jinja2==2.9.6,Mako==1.0.6,MarkupSafe==1.0,parse==1.8.2,parse-type==0.3.4,py==1.4.34,Pygments==2.2.0,pyPEG2==2.15.2,pytest==3.1.1,pytest-bdd==2.18.2,pytest-benchmark==3.0.0,pytest-catchlog==1.2.2,pytest-cov==2.5.1,pytest-faulthandler==1.3.1,pytest-instafail==0.3.0,pytest-mock==1.6.0,pytest-qt==2.1.0,pytest-repeat==0.4.1,pytest-rerunfailures==2.1.0,pytest-travis-fold==1.2.0,pytest-xvfb==1.0.0,PyVirtualDisplay==0.2.1,PyYAML==3.12,six==1.10.0,vulture==0.14,Werkzeug==0.12.2
    py34 runtests: PYTHONHASHSEED='3683105317'
    py34 runtests: commands[0] | /home/caveman/Documents/dev/qutebrowser/.tox/py34/bin/python scripts/link_pyqt.py --tox /home/caveman/Documents/dev/qutebrowser/.tox/py34
    py34 runtests: commands[1] | /home/caveman/Documents/dev/qutebrowser/.tox/py34/bin/python -bb -m pytest tests/unit/utils/test_utils.py
    INTERNALERROR> Traceback (most recent call last):
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/_pytest/main.py", line 101, in wrap_session
    INTERNALERROR>     config._do_configure()
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/_pytest/config.py", line 906, in _do_configure
    INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 750, in call_historic
    INTERNALERROR>     self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    INTERNALERROR>     res = hook_impl.function(*args)
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/pytestqt/plugin.py", line 167, in pytest_configure
    INTERNALERROR>     qt_api.set_qt_api(config.getini('qt_api'))
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/pytestqt/qt_compat.py", line 85, in set_qt_api
    INTERNALERROR>     self.QtTest = _import_module('QtTest')
    INTERNALERROR>   File "/home/caveman/Documents/dev/qutebrowser/.tox/py34/lib/python3.4/site-packages/pytestqt/qt_compat.py", line 69, in _import_module
    INTERNALERROR>     return getattr(m, module_name)
    INTERNALERROR> AttributeError: 'module' object has no attribute 'QtTest'
    ERROR: InvocationError: '/home/caveman/Documents/dev/qutebrowser/.tox/py34/bin/python -bb -m pytest tests/unit/utils/test_utils.py'
    _____________________________________ summary ______________________________________
    ERROR:   py34: commands failed

Sorry, this is my 1st test ever.

Al-Caveman commented:
Just to summarize everything:

 1. The feature is complete and functional as far as I can see.
 2. The first version of the test of 2 corresponding `utils.py` functions is added.
 3. I don't know how to run the tests (I just added the tests). So I'm not sure if my tests actually work.
 4. I'm considering transforming `indexset_p` into an object (by defining a class `IndexSet` in `utils.py` as suggested by @Kingdread). Assuming this goes smoothly, I'm also considering finding a better name for it (maybe `NumbersSet`?).

Kingdread commented:
`tox -e py34 -- tests/unit/utils/test_utils.py` should work, there are probably some outdated dependencies. You can try recreating the environment (`tox -re py34`), which updates the dependencies.

The-Compiler commented:
Closing this as abandoned since I haven't heard back in a while now. I added the patch to #2384 though, to make sure nothing gets lost. Thanks for you contribution anyways, @Al-Caveman!

Al-Caveman commented:
Sorry was too busy lately. I managed some time now, and I think I addressed all concerns (commit linked in #2384):

  1. My functions are in a new class named `NumberSet` that supports `in` operator.
  2. `downloads.py` is updated accordingly.
  3. Tests in `test_utils.py` is complete as far as I know.

While I didn't use `tox` to run the tests, I did run my test code in `test_utils.py` in my own environment, and it all worked successfully. So, in *theory*, I think my `test_utils.py` code is perfect.
